packages/material-ui-system/src/styleFunctionSx.js

```diff
@@ -23,13 +23,21 @@ const filterProps = [
   ...sizing.filterProps,
   ...spacing.filterProps,
   ...typography.filterProps,
-  'sx',
 ];
 
 function objectsHaveSameKeys(...objects) {
-  const allKeys = objects.reduce((keys, object) => keys.concat(Object.keys(object)), []);
+  const objectsKeysLength = [];
+
+  const allKeys = objects.reduce((keys, object) => {
+    const objectKeys = Object.keys(object);
+
+    objectsKeysLength.push(objectKeys.length);
+    return keys.concat(objectKeys);
+  }, []);
+
   const union = new Set(allKeys);
-  return objects.every((object) => union.size === Object.keys(object).length);
```

Comment:
Changes are done in order to avoid invoking `Object.keys(object)` more than once on the same object.